src/main/haskell/kore/src/Kore/AST/Kore.hs

```diff
@@ -154,62 +182,141 @@ deriving instance
+
+@KorePattern@ covers the 'Object' and 'Meta' levels of Kore, corresponding to
+the syntactic category @pattern@ in The Semantics of K, Section 9.1.4
+(Patterns).
+
+@dom@ is the type of domain values; see "Kore.Domain.External" and
+"Kore.Domain.Builtin".
+
+@var@ is the family of variable types, parameterized by level.
+
+@ann@ is the type of annotations decorating each node of the abstract syntax
+tree. @KorePattern@ is a 'Traversable' 'Comonad' over the type of annotations.
+
+-}
+newtype KorePattern
+    (dom :: * -> *)
+    (var :: * -> *)
+    (ann :: *)
+  =
+    KorePattern { getKorePattern :: Cofree (UnifiedPattern dom var) ann }
+    deriving (Foldable, Functor, Generic, Traversable)
+
+instance
+    ( Eq ann
+    , EqMetaOrObject var
+    , Eq1 dom, Functor dom
+    ) =>
+    Eq (KorePattern dom var ann)
+  where
+    (==) = eqWorker
+      where
+        eqWorker
+            (Recursive.project -> ann1 :< pat1)
+            (Recursive.project -> ann2 :< pat2)
+          =
+            ann1 == ann2 && liftEq eqWorker pat1 pat2
+
+instance
+    ( Ord ann
+    , OrdMetaOrObject var
+    , Ord1 dom, Functor dom
+    ) =>
+    Ord (KorePattern dom var ann)
+  where
+    compare = compareWorker
+      where
+        compareWorker
+            (Recursive.project -> ann1 :< pat1)
+            (Recursive.project -> ann2 :< pat2)
+          =
+            compare ann1 ann2 <> liftCompare compareWorker pat1 pat2
+
+deriving instance
+    ( Show ann
+    , ShowMetaOrObject var
+    , Show (dom child)
+    , child ~ Cofree (UnifiedPattern dom var) ann
+    ) =>
+    Show (KorePattern dom var ann)
+
+instance
+    ( Functor dom
+    , Hashable ann
+    , Hashable (var Meta)
+    , Hashable (var Object)
+    , Hashable (dom child)
+    , child ~ KorePattern dom var ann
+    ) =>
+    Hashable (KorePattern dom var ann)
+  where
+    hashWithSalt salt (Recursive.project -> ann :< pat) =
+        salt `hashWithSalt` ann `hashWithSalt` pat
 
-pattern KoreObjectPattern
-    :: Pattern Object domain var (KorePattern domain var)
-    -> KorePattern domain var
-pattern KoreObjectPattern pat = Fix (UnifiedObjectPattern pat)
+instance
+    ( Functor dom
+    , NFData ann
+    , NFData (var Meta)
+    , NFData (var Object)
+    , NFData (dom child)
+    , child ~ KorePattern dom var ann
+    ) =>
+    NFData (KorePattern dom var ann)
+  where
+    rnf (Recursive.project -> ann :< pat) =
+        rnf ann `seq` rnf pat `seq` ()
+
+type instance Base (KorePattern dom var ann) =
+    CofreeF (UnifiedPattern dom var) ann
+
+instance Functor dom => Recursive (KorePattern dom var ann) where
+    project (KorePattern embedded) =
+        case Recursive.project embedded of
+            Compose (Identity projected) -> KorePattern <$> projected
 
-{-# COMPLETE KoreMetaPattern, KoreObjectPattern #-}
+instance Functor dom => Corecursive (KorePattern dom var ann) where
+    embed projected =
+        (KorePattern . Recursive.embed . Compose . Identity)
```

Comment:
The [`Base`](https://hackage.haskell.org/package/recursion-schemes-5.0.3/docs/Data-Functor-Foldable.html#t:Base) instance for [`Cofree`](https://hackage.haskell.org/package/free-5.0.2/docs/Control-Comonad-Trans-Cofree.html#t:Cofree) is defined in terms of those `newtype` wrappers. The `Recursive` and `Corecursive` instances of `PurePattern` and `KorePattern` are the only places they should appear in our code because these instances will handle all the wrapping and unwrapping we need to do.